course-content/34_machine_learning.qmd

```diff
@@ -0,0 +1,244 @@
+
+```{python}
+X_train.shape, y_train.shape, X_test.shape, y_test.shape
+```
+
+The training objects have 144 rows of the original dataframe and the test objects have 100 rows.  The X objects have 11 feature variables and the y objects consist of one-dimensional arrays, which show  as "blank" to the right of the comma when reviewing object `shape`.
+
+
+## Fitting and Predicting 
+
+With these preprocessing steps now complete we can apply a classifier method to the training data. Scikit-learn has many classifier methods. For our example, we will use a machine learning algorithm called `LogisticRegression()`. Despite it having "regression" in the name, it is implemented here for classification rather than regression. In logistic regression, the dependent variable is a binary variable that contains data coded as 1 (yes, success, etc.) or 0 (no, failure, etc.). 
+
+In our code, we first call `LogisticRegression()` and then `fit()` method then attempts to fit the independent variables to the target variable in the training dataset. In the code below, we run the fitting process and then take that fitted algorithm to product predictions for our test data. 
+
+
+```{python}
+from sklearn.linear_model import LogisticRegression
+
+reg = LogisticRegression()
+reg.fit(X_train, y_train)
+```
+<br>
+Now our we have a trained model and we can run that model on our test dataset. We enter X_test as an argument following the `predict()` and `predict_proba()` methods as in the code below. Let us look at the first five cases in our array of predictions:
+
+```{python}
+y_predict =  reg.predict(X_test) # makes a decision based on probability
+y_predict[:5] # first 5 prediction decisions in test dataset
+```
+The output gives us an array of predictions for the first five cases in sequence in the test dataset. A value of 1 means that the model predicts the party to be a big tipper, and 0 predicts that it will not be a big tipper. 
+
+Each of the predictions is based on a discrete probability. The `predict_proba()` method returns the class probabilities for each data point.  The number of probabilities for each row is equal to the number of categories in target variable (2 in our case).
+
+```{python}
+y_predict_proba =  reg.predict_proba(X_test) # computes a probability per case
+y_predict_proba[:5] # first 5 predictions in test dataset
+```
+
+We can see for each case a pair of probabilities, one for "0" and the other for "1". This detail helps us understand how the decisions with the `predict()` method were made. 
+
+
+## Evaluating
+
+Ok. We have our predictions... but are they good? Of course, what "good" means needs a bit of thought. There are many performance measures in data science and there are considerations based on the use cases themselves. How we would evaluate performace of an algorithm that attempts to diagnose serious disease would be different than if we were evaluating one that identified whether an email was spam or not. 
```

Comment:
```suggestion
Ok. We have our predictions... but are they good? Of course, what "good" means needs a bit of thought. There are many performance measures in data science and there are considerations based on the use cases themselves. How we would evaluate performance of an algorithm that attempts to diagnose serious disease would be different than if we were evaluating one that identified whether an email was spam or not. In the case of disease classification, we would not want a high false negative rate, and thus missing out on important treatment for people who actually have the disease. On the other hand with spam filters, it would be okay to miss some spam, but it would be worse to have a high false positive rate, and start chucking out real emails with possible spam. 
```

Reply:
I reworked this content into the discussion of precision and recall